Sources/EdgeNetworkHandlers/EdgeHitProcessor.swift

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright 2020 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may obtain a copy
+// of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+import AEPCore
+import AEPServices
+import Foundation
+
+/// A `HitProcessing` which handles the processing of `EdgeHit`s
+class EdgeHitProcessor: HitProcessing {
+    private let LOG_TAG = "EdgeHitProcessor"
+    private var networkService: EdgeNetworkService
+    private var networkResponseHandler: NetworkResponseHandler
+    private var getSharedState: (String, Event?) -> SharedStateResult?
+    let retryInterval = Constants.NetworkKeys.RETRY_INTERVAL
+
+    init(networkService: EdgeNetworkService, networkResponseHandler: NetworkResponseHandler, getSharedState: @escaping (String, Event?) -> SharedStateResult?) {
+        self.networkService = networkService
+        self.networkResponseHandler = networkResponseHandler
+        self.getSharedState = getSharedState
+    }
+
+    func processHit(entity: DataEntity, completion: @escaping (Bool) -> Void) {
+        guard let data = entity.data, let event = try? JSONDecoder().decode(Event.self, from: data) else {
+            // can't convert data to hit, unrecoverable error, move to next hit
+            Log.debug(label: LOG_TAG, "processHit - Failed to decode edge hit with id '\(entity.uniqueIdentifier)'.")
+            completion(true)
+            return
+        }
+
+        // fetch config shared state, this should be resolved based on readyForEvent check
+        guard let configId = getEdgeConfigId(event: event) else {
+            Log.debug(label: LOG_TAG, "processHit - Failed get edge config id '\(entity.uniqueIdentifier)'.")
+            completion(true)
+            return // drop current event
+        }
+
+        // Build Request object
+        let requestBuilder = RequestBuilder()
+        requestBuilder.enableResponseStreaming(recordSeparator: Constants.Defaults.RECORD_SEPARATOR,
+                                               lineFeed: Constants.Defaults.LINE_FEED)
+
+        // get ECID from Identity shared state, this should be resolved based on readyForEvent check
+        guard let identityState =
+                getSharedState(Constants.SharedState.Identity.STATE_OWNER_NAME,
+                               event)?.value else {
+            Log.warning(label: LOG_TAG,
+                        "processHit - Unable to process the event '\(event.id.uuidString)', " +
+                            "Identity shared state is nil.")
+            completion(true)
```

Comment:
Is it possible for a hit to be processed before the Identity extension is fully initialized?  As the hit queue is setup and started from the Edge.init() ctor, I'm wondering, if an app is launched with persisted hits, could they get processed before Identity was able to save its first state? Since Edge is configured to handle events only when Configuration and Identity have shared states, maybe the same should be true here?

Reply:
Yeah, this is a good catch, I see a few options:

1. Only start hit processing once `readyForEvent` would return true to indicate that all the required shared states are published.
2. Change line 58 in EdgeHitProcessor to `completion(false)` so that this hit will be retried until Identity has provided a shared state.

I'm leaning towards option 2. Thoughts?